src/nsolid/nsolid_heap_snapshot.cc

```diff
@@ -12,6 +14,59 @@ NSolidHeapSnapshot::NSolidHeapSnapshot() {
   ASSERT_EQ(0, in_progress_heap_snapshots_.init(true));
 }
 
+int NSolidHeapSnapshot::StartTrackingHeapObjects(
+    SharedEnvInst envinst,
+    bool redacted,
+    bool trackAllocations,
+    uint64_t duration,
+    void* data,
+    Snapshot::snapshot_proxy_sig proxy,
+    internal::deleter_sig deleter) {
+  uint64_t thread_id = envinst->thread_id();
+  nsuv::ns_mutex::scoped_lock lock(&in_progress_heap_snapshots_);
+  // We can not trigger this command if there is already a snapshot in progress
+  auto it = threads_running_snapshots_.find(thread_id);
+  if (it != threads_running_snapshots_.end()) {
+    return UV_EEXIST;
+  }
+
+  int status = RunCommand(envinst,
+                          CommandType::Interrupt,
+                          start_tracking_heapobjects,
+                          trackAllocations,
+                          duration,
+                          this);
+
+  // Consider this as taking a heap snapshot in the thread
+  if (status == 0) {
+    // Now we are tracking heap objects in this thread
+    threads_running_snapshots_.emplace(
+        thread_id,
+        HeapSnapshotStor{
+            redacted, true, proxy, internal::user_data(data, deleter)});
+  }
+
+  return status;
```

Comment:
We can do a bit better:

```suggestion
  auto it = threads_running_snapshots_.emplace(
        thread_id,
        HeapSnapshotStor{
            redacted, true, proxy, internal::user_data(data, deleter)});

  if (it.second == false) {
    return UV_EEXIST;
  }

  int status = RunCommand(envinst,
                          CommandType::Interrupt,
                          start_tracking_heapobjects,
                          trackAllocations,
                          duration,
                          this);

  // Consider this as taking a heap snapshot in the thread
  if (status == 0) {
    // Now we are tracking heap objects in this thread
    threads_running_snapshots_.erase(it.first);
  }

  return status;
```

Reply:
Thanks! Will fix that!

Reply:
This hasn't been resolved afaict

Reply:
This suggestion provokes the runtime to abort. I'll proceed to resolve this conversation.

```lldb
(lldb) bt all
* thread #1, queue = 'com.apple.main-thread', stop reason = signal SIGABRT
  * frame #0: 0x000000018fee0180 libsystem_kernel.dylib`__kill + 8
    frame #1: 0x00000001015c78e8 nsolid`uv_kill(pid=0, signum=6) at process.c:1091:7
    frame #2: 0x00000001002edc78 nsolid`node::Kill(args=0x000000016fdfcfe8) at node_process_methods.cc:166:13
    frame #3: 0x0000000100788d8c nsolid`v8::internal::FunctionCallbackArguments::Call(this=0x000000016fdfd088, handler=<unavailable>) at api-arguments-inl.h:146:3 [opt]
    frame #4: 0x0000000100787af8 nsolid`v8::internal::MaybeHandle<v8::internal::Object> v8::internal::(anonymous namespace)::HandleApiCallHelper<false>(isolate=0x0000000110008000, new_target=Handle<v8::internal::HeapObject> @ x23, fun_data=<unavailable>, receiver=<unavailable>, argv=0x000000016fdfd228, argc=2) at builtins-api.cc:113:36 [opt]
    frame #5: 0x00000001007866e4 nsolid`v8::internal::Builtin_HandleApiCall(int, unsigned long*, v8::internal::Isolate*) [inlined] v8::internal::Builtin_Impl_HandleApiCall(args=<unavailable>, isolate=0x0000000110008000) at builtins-api.cc:144:5 [opt]
    frame #6: 0x00000001007865fc nsolid`v8::internal::Builtin_HandleApiCall(args_length=<unavailable>, args_object=<unavailable>, isolate=0x0000000110008000) at builtins-api.cc:135:1 [opt]
    frame #7: 0x0000000101b1e4a4 nsolid`Builtins_CEntry_Return1_ArgvOnStack_BuiltinExit + 100
    frame #8: 0x0000000101a83404 nsolid`Builtins_InterpreterEntryTrampoline + 260
    frame #9: 0x0000000101a83404 nsolid`Builtins_InterpreterEntryTrampoline + 260
    frame #10: 0x0000000101a83404 nsolid`Builtins_InterpreterEntryTrampoline + 260
    frame #11: 0x0000000101a83404 nsolid`Builtins_InterpreterEntryTrampoline + 260
    frame #12: 0x0000000101a83404 nsolid`Builtins_InterpreterEntryTrampoline + 260
    frame #13: 0x0000000101a83404 nsolid`Builtins_InterpreterEntryTrampoline + 260
    frame #14: 0x0000000101a83404 nsolid`Builtins_InterpreterEntryTrampoline + 260
    frame #15: 0x0000000101a83404 nsolid`Builtins_InterpreterEntryTrampoline + 260
    frame #16: 0x0000000101a83404 nsolid`Builtins_InterpreterEntryTrampoline + 260
    frame #17: 0x0000000101a83404 nsolid`Builtins_InterpreterEntryTrampoline + 260
    frame #18: 0x0000000101a83404 nsolid`Builtins_InterpreterEntryTrampoline + 260
    frame #19: 0x0000000101a83404 nsolid`Builtins_InterpreterEntryTrampoline + 260
    frame #20: 0x0000000101a83404 nsolid`Builtins_InterpreterEntryTrampoline + 260
    frame #21: 0x0000000101a83404 nsolid`Builtins_InterpreterEntryTrampoline + 260
    frame #22: 0x0000000101a83404 nsolid`Builtins_InterpreterEntryTrampoline + 260
    frame #23: 0x0000000101a80dec nsolid`Builtins_JSEntryTrampoline + 172
    frame #24: 0x0000000101a80ad4 nsolid`Builtins_JSEntry + 148
    frame #25: 0x000000010097cd60 nsolid`v8::internal::(anonymous namespace)::Invoke(v8::internal::Isolate*, v8::internal::(anonymous namespace)::InvokeParams const&) [inlined] v8::internal::GeneratedCode<unsigned long, unsigned long, unsigned long, unsigned long, unsigned long, long, unsigned long**>::Call(this=<unavailable>, args=<unavailable>, args=<unavailable>, args=<unavailable>, args=<unavailable>, args=<unavailable>, args=<unavailable>) at simulator.h:154:12 [opt]
    frame #26: 0x000000010097cd5c nsolid`v8::internal::(anonymous namespace)::Invoke(isolate=0x0000000110008000, params=0x000000016fdfe1b8)::InvokeParams const&) at execution.cc:427:33 [opt]
    frame #27: 0x000000010097bf48 nsolid`v8::internal::Execution::Call(isolate=0x0000000110008000, callable=Handle<v8::internal::Object> @ x23, receiver=Handle<v8::internal::Object> @ x22, argc=4, argv=0x00006000002291a0) at execution.cc:529:10 [opt]
    frame #28: 0x00000001006db5a8 nsolid`v8::Function::Call(this=<unavailable>, context=<unavailable>, recv=(val_ = 0x0000000110008210), argc=4, argv=0x00006000002291a0) at api.cc:5547:7 [opt]
    frame #29: 0x00000001001a74e8 nsolid`node::builtins::BuiltinLoader::CompileAndCall(this=0x000000010a82fa60, context=(val_ = 0x000000010a87eb00), id="internal/main/run_main_module", argc=4, argv=0x00006000002291a0, optional_realm=0x000000010b00ac60) at node_builtins.cc:482:14
    frame #30: 0x00000001001a736c nsolid`node::builtins::BuiltinLoader::CompileAndCall(this=0x000000010a82fa60, context=(val_ = 0x000000010a87eb00), id="internal/main/run_main_module", realm=0x000000010b00ac60) at node_builtins.cc:466:10
    frame #31: 0x00000001002fd3e0 nsolid`node::Realm::ExecuteBootstrapper(this=0x000000010b00ac60, id="internal/main/run_main_module") at node_realm.cc:161:32
    frame #32: 0x0000000100164604 nsolid`node::StartExecution(env=0x000000010a82f000, main_script_id="internal/main/run_main_module") at node.cc:277:35
    frame #33: 0x000000010016440c nsolid`node::StartExecution(env=0x000000010a82f000, cb=node::StartExecutionCallback @ 0x000000016fdfe938)>) at node.cc:366:12
    frame #34: 0x00000001000156a0 nsolid`node::LoadEnvironment(env=0x000000010a82f000, cb=node::StartExecutionCallback @ 0x000000016fdfea48)>) at environment.cc:549:10
    frame #35: 0x0000000100270304 nsolid`node::NodeMainInstance::Run(this=0x000000016fdfeb80, exit_code=0x000000016fdfead4, env=0x000000010a82f000) at node_main_instance.cc:110:7
    frame #36: 0x000000010026ff50 nsolid`node::NodeMainInstance::Run(this=0x000000016fdfeb80) at node_main_instance.cc:90:3
    frame #37: 0x0000000100166e20 nsolid`node::StartInternal(argc=2, argv=0x00006000030043f0) at node.cc:1397:24
    frame #38: 0x0000000100166a68 nsolid`node::Start(argc=2, argv=0x000000016fdfef98) at node.cc:1404:27
    frame #39: 0x00000001020141c4 nsolid`main(argc=2, argv=0x000000016fdfef98) at node_main.cc:97:10
    frame #40: 0x000000018fbbff28 dyld`start + 2236
  thread #2
    frame #0: 0x000000018fede0a0 libsystem_kernel.dylib`kevent + 8
    frame #1: 0x00000001015dab58 nsolid`uv__io_poll(loop=0x000000010a81b8e8, timeout=-1) at kqueue.c:257:12
    frame #2: 0x00000001015b8338 nsolid`uv_run(loop=0x000000010a81b8e8, mode=UV_RUN_DEFAULT) at core.c:453:5
    frame #3: 0x00000001002d4b88 nsolid`node::WorkerThreadsTaskRunner::DelayedTaskScheduler::Run(this=0x000000010a81b800) at node_platform.cc:95:5
    frame #4: 0x00000001002d4a20 nsolid`node::WorkerThreadsTaskRunner::DelayedTaskScheduler::Start(this=0x000000010a81b800, data=0x000000010a81b800)::'lambda'(void*)::operator()(void*) const at node_platform.cc:64:49
    frame #5: 0x00000001002d49f8 nsolid`node::WorkerThreadsTaskRunner::DelayedTaskScheduler::Start(data=0x000000010a81b800)::'lambda'(void*)::__invoke(void*) at node_platform.cc:63:25
    frame #6: 0x000000018ff17fa8 libsystem_pthread.dylib`_pthread_start + 148
  thread #3
    frame #0: 0x000000018fedb750 libsystem_kernel.dylib`__psynch_cvwait + 8
    frame #1: 0x000000018ff18574 libsystem_pthread.dylib`_pthread_cond_wait + 1232
    frame #2: 0x00000001015d29ec nsolid`uv_cond_wait(cond=0x000000010b0064c8, mutex=0x000000010b006488) at thread.c:793:7
    frame #3: 0x000000010017d57c nsolid`node::LibuvMutexTraits::cond_wait(cond=0x000000010b0064c8, mutex=0x000000010b006488) at node_mutex.h:175:5
    frame #4: 0x000000010017d524 nsolid`node::ConditionVariableBase<node::LibuvMutexTraits>::Wait(this=0x000000010b0064c8, scoped_lock=0x0000000170e0af38) at node_mutex.h:249:3
    frame #5: 0x00000001002d5334 nsolid`node::TaskQueue<v8::Task>::BlockingPop(this=0x000000010b006488) at node_platform.cc:600:22
    frame #6: 0x00000001002d16ac nsolid`node::(anonymous namespace)::PlatformWorkerThread(data=0x0000600000c08000) at node_platform.cc:42:61
    frame #7: 0x000000018ff17fa8 libsystem_pthread.dylib`_pthread_start + 148
  thread #4
    frame #0: 0x000000018fedb750 libsystem_kernel.dylib`__psynch_cvwait + 8
    frame #1: 0x000000018ff18574 libsystem_pthread.dylib`_pthread_cond_wait + 1232
    frame #2: 0x00000001015d29ec nsolid`uv_cond_wait(cond=0x000000010b0064c8, mutex=0x000000010b006488) at thread.c:793:7
    frame #3: 0x000000010017d57c nsolid`node::LibuvMutexTraits::cond_wait(cond=0x000000010b0064c8, mutex=0x000000010b006488) at node_mutex.h:175:5
    frame #4: 0x000000010017d524 nsolid`node::ConditionVariableBase<node::LibuvMutexTraits>::Wait(this=0x000000010b0064c8, scoped_lock=0x0000000171612f38) at node_mutex.h:249:3
    frame #5: 0x00000001002d5334 nsolid`node::TaskQueue<v8::Task>::BlockingPop(this=0x000000010b006488) at node_platform.cc:600:22
    frame #6: 0x00000001002d16ac nsolid`node::(anonymous namespace)::PlatformWorkerThread(data=0x0000600000c08030) at node_platform.cc:42:61
    frame #7: 0x000000018ff17fa8 libsystem_pthread.dylib`_pthread_start + 148
  thread #5
    frame #0: 0x000000018fedb750 libsystem_kernel.dylib`__psynch_cvwait + 8
    frame #1: 0x000000018ff18574 libsystem_pthread.dylib`_pthread_cond_wait + 1232
    frame #2: 0x00000001015d29ec nsolid`uv_cond_wait(cond=0x000000010b0064c8, mutex=0x000000010b006488) at thread.c:793:7
    frame #3: 0x000000010017d57c nsolid`node::LibuvMutexTraits::cond_wait(cond=0x000000010b0064c8, mutex=0x000000010b006488) at node_mutex.h:175:5
    frame #4: 0x000000010017d524 nsolid`node::ConditionVariableBase<node::LibuvMutexTraits>::Wait(this=0x000000010b0064c8, scoped_lock=0x0000000171e1af38) at node_mutex.h:249:3
    frame #5: 0x00000001002d5334 nsolid`node::TaskQueue<v8::Task>::BlockingPop(this=0x000000010b006488) at node_platform.cc:600:22
    frame #6: 0x00000001002d16ac nsolid`node::(anonymous namespace)::PlatformWorkerThread(data=0x0000600000c08060) at node_platform.cc:42:61
    frame #7: 0x000000018ff17fa8 libsystem_pthread.dylib`_pthread_start + 148
  thread #6
    frame #0: 0x000000018fedb750 libsystem_kernel.dylib`__psynch_cvwait + 8
    frame #1: 0x000000018ff18574 libsystem_pthread.dylib`_pthread_cond_wait + 1232
    frame #2: 0x00000001015d29ec nsolid`uv_cond_wait(cond=0x000000010b0064c8, mutex=0x000000010b006488) at thread.c:793:7
    frame #3: 0x000000010017d57c nsolid`node::LibuvMutexTraits::cond_wait(cond=0x000000010b0064c8, mutex=0x000000010b006488) at node_mutex.h:175:5
    frame #4: 0x000000010017d524 nsolid`node::ConditionVariableBase<node::LibuvMutexTraits>::Wait(this=0x000000010b0064c8, scoped_lock=0x0000000172622f38) at node_mutex.h:249:3
    frame #5: 0x00000001002d5334 nsolid`node::TaskQueue<v8::Task>::BlockingPop(this=0x000000010b006488) at node_platform.cc:600:22
    frame #6: 0x00000001002d16ac nsolid`node::(anonymous namespace)::PlatformWorkerThread(data=0x0000600000c08090) at node_platform.cc:42:61
    frame #7: 0x000000018ff17fa8 libsystem_pthread.dylib`_pthread_start + 148
  thread #7
    frame #0: 0x000000018fede0a0 libsystem_kernel.dylib`kevent + 8
    frame #1: 0x00000001015dab58 nsolid`uv__io_poll(loop=0x0000000105943bf8, timeout=100) at kqueue.c:257:12
    frame #2: 0x00000001015b8338 nsolid`uv_run(loop=0x0000000105943bf8, mode=UV_RUN_DEFAULT) at core.c:453:5
    frame #3: 0x00000001004f46b8 nsolid`node::nsolid::EnvList::env_list_routine_((null)=0x0000000105944168, envlist=0x0000000105943bd8) at nsolid_api.cc:1490:8
    frame #4: 0x00000001005161ec nsolid`void nsuv::ns_thread::create_proxy_<void (*)(nsuv::ns_thread*, node::nsolid::EnvList*), node::nsolid::EnvList>(arg=0x0000000105944168) at nsuv-inl.h:1829:3
    frame #5: 0x000000018ff17fa8 libsystem_pthread.dylib`_pthread_start + 148
  thread #8
    frame #0: 0x000000018fed7ed0 libsystem_kernel.dylib`semaphore_wait_trap + 8
    frame #1: 0x00000001015d2744 nsolid`uv_sem_wait(sem=0x0000000105944cec) at thread.c:471:9
    frame #2: 0x00000001005517c0 nsolid`node::inspector::(anonymous namespace)::StartIoThreadMain(unused=0x0000000000000000) at inspector_agent.cc:85:5
    frame #3: 0x000000018ff17fa8 libsystem_pthread.dylib`_pthread_start + 148

```